master/buildbot/reporters/slack.py

```diff
@@ -0,0 +1,128 @@
+from buildbot.util import httpclientservice
+from buildbot.reporters.http import HttpStatusPushBase
+
+SLACK_BASE_URL = "https://hooks.slack.com"
+
+
+class SlackStatusPush(HttpStatusPushBase):
+    """
+    Sends messages to a Slack.io channel when each build finishes with a handy
+    link to the build results.
+    """
+
+    name = "SlackStatusPush"
+
+    def checkConfig(self, endpoint, username="buildbot",
+                    icon="https://buildbot.net/img/nut.png",
+                    notify_on_success=True, notify_on_failure=True,
+                    **kwargs):
+        if not isinstance(endpoint, string_types):
+            config.error('endpoint must be a string')
+        if not isinstance(username, string_types):
+            config.error('username must be a string')
+        if not isinstance(icon, string_types):
+            config.error('icon must be a string')
+        if not isinstance(notify_on_success, bool):
+            config.error('notify_on_success must be a bool')
+        if not isinstance(notify_on_failure, bool):
+            config.error('notify_on_failure must be a bool')
+
+    @defer.inlineCallbacks
+    def reconfigService(self, endpoint, username="buildbot",
+                        icon="https://buildbot.net/img/nut.png",
+                        notify_on_success=True, notify_on_failure=True,
+                        **kwargs):
+        """
+        Creates a SlackStatusPush status service.
+        :param endpoint: Your Slack endpoint.
+        :param username: The user name of the "user" positing the messages on
+            Slack.
+        :param icon: The icon of the user posting the messages on Slack.
+        :param notify_on_success: Set this to False if you don't want
+            messages when a build was successful.
+        :param notify_on_failure: Set this to False if you don't want
+            messages when a build failed.
+        """
+        yield HttpStatusPushBase.reconfigService(self, **kwargs)
+        self._http = yield httpclientservice.HTTPClientService.getService(
+            self.master, SLACK_BASE_URL,
+            debug=self.debug, verify=self.verify)
+
+        self.endpoint = endpoint
+        self.username = username
+        self.icon = icon
+        self.notify_on_success = notify_on_success
+        self.notify_on_failure = notify_on_failure
+
+    @defer.inlineCallbacks
+    def buildFinished(self, key, build):
+        if not self.notify_on_success and build['result'] == SUCCESS:
+            return
+
+        if not self.notify_on_failure and build['result'] != SUCCESS:
+            return
+
+        url = "%s#/builders/%d/builds/%d" % (self.master.config.buildbotURL,
+                                             build['builderid'],
+                                             build['buildid'])
+        naive_startedAt = build['started_at'].replace(tzinfo=None)
+        humantime = naturaltime(datetime.now() - naive_startedAt)
```

Comment:
why don't use complete_at?

Reply:
Wouldn't `complete_at` always give `1 second ago`, since the notification is pushed immediately after completion? We want to know when the request was submitted.

Reply:
complete_at - started_at!